File: backend/app/schemas/movimentos.py
"""
FORMATO DOS PEDIDOS E RESPOSTAS DA API DE MOVIMENTOS
=====================================================

Estas classes (baseadas em BaseModel, do Pydantic) descrevem o formato dos
dados que entram e saem dos endpoints de movimentos — são diferentes do
modelo em app/models/movimento.py, que descreve a tabela da base de
dados.

VALOR COM SINAL: um único campo, positivo para uma entrada e negativo para
uma saída (ver a nota em app/models/movimento.py). O ecrã de
criar/editar um movimento pode continuar a perguntar "entrada ou saída?" e
pedir um valor sempre positivo — é só a forma de apresentação; a conversão
para um valor com sinal faz-se no próprio formulário, antes do pedido
chegar aqui.

CATEGORIA OBRIGATÓRIA: categoria_id é um campo próprio, não opcional — ver
a nota CATEGORIA OBRIGATÓRIA em app/models/movimento.py. A rota
(app/routers/movimentos.py) confirma que a categoria indicada pertence ao
utilizador e que a sua direcao (entrada/saida) é coerente com o sinal do
valor — tal como para conta_id, este schema só valida a FORMA dos dados.

Os campos são os mesmos na criação e na edição (ao contrário de Conta, uma
edição de movimento pode mexer em tudo, incluindo mudar a conta a que
pertence — "mover" o movimento) — por isso vivem numa única classe-base
(_CamposMovimento), sem precisar de duas versões.
"""

import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, Field, field_validator


class _CamposMovimento(BaseModel):
    """Os campos de um movimento, comuns à criação e à edição."""

    # A conta a que este movimento pertence. Tem de pertencer ao
    # utilizador autenticado — verificado na rota (app/routers/
    # movimentos.py), não aqui: este schema só sabe validar a FORMA dos
    # dados, não se um determinado id existe ou é deste utilizador.
    conta_id: uuid.UUID

    # A categoria atribuída a este movimento. Tem de pertencer ao
    # utilizador autenticado, e a sua direcao tem de corresponder ao sinal
    # de "valor" (uma categoria de entrada não serve para uma saída) —
    # verificado na rota, pela mesma razão de conta_id: este schema só
    # valida a forma dos dados, não se um determinado id existe, é deste
    # utilizador, ou é coerente com o resto do pedido.
    categoria_id: uuid.UUID

    data: date

    descricao: str = Field(min_length=1, max_length=200)

    # Com sinal: positivo é uma entrada, negativo é uma saída.
    valor: Decimal

    @field_validator("descricao")
    @classmethod
    def _descricao_sem_espacos_e_nao_vazia(cls, valor: str) -> str:
        valor = valor.strip()
        if not valor:
            raise ValueError("A descrição não pode ser vazia.")
        return valor

    @field_validator("valor")
    @classmethod
    def _valor_nao_pode_ser_zero(cls, valor: Decimal) -> Decimal:
        # Um movimento de valor 0 não muda saldo nenhum — não corresponde
        # a nenhuma transacção real, é quase de certeza um erro de
        # preenchimento.
        if valor == 0:
            raise ValueError("O valor não pode ser zero.")
        return valor


class MovimentoCriar(_CamposMovimento):
    """Dados recebidos no pedido de criação de um movimento (POST /movimentos)."""


class MovimentoEditar(_CamposMovimento):
    """Dados recebidos no pedido de edição de um movimento (PATCH /movimentos/{id})."""


class MovimentoOut(BaseModel):
    """Dados de um movimento devolvidos pela API."""

    id: uuid.UUID
    conta_id: uuid.UUID
    categoria_id: uuid.UUID
    data: date
    descricao: str

    # Texto, não número (ver a mesma nota em app/schemas/contas.py): o
    # JSON não tem um tipo decimal exacto.
    valor: str

    created_at: datetime
    updated_at: datetime

File: frontend/src/lib/movimentos.ts
/*
 * CLIENTE HTTP DA API DE MOVIMENTOS
 * =================================
 *
 * As funções que falam com os endpoints /movimentos do backend. O resto
 * da aplicação (as páginas de movimentos) chama estas funções e nunca usa
 * "fetch" diretamente. A mecânica comum (prefixo, cookie, erros) vive em
 * src/lib/http.ts — o mesmo padrão de src/lib/contas.ts.
 */

import { pedido } from './http'

/**
 * Um movimento, tal como a API o devolve (schema MovimentoOut do
 * backend).
 *
 * Notas:
 *   - "valor" é texto ("1234.56" ou "-50.00"), não número — o JSON não
 *     tem tipo decimal e um número perderia precisão em cêntimos.
 *   - COM SINAL: positivo é uma entrada, negativo é uma saída. O ecrã de
 *     criar/editar pergunta "entrada ou saída?" e pede um valor sempre
 *     positivo — a conversão para um valor com sinal faz-se no próprio
 *     formulário, antes de chamar criarMovimento/editarMovimento.
 *   - "data" / created_at / updated_at são strings ISO.
 *   - "categoria_id" é sempre obrigatório — nunca "sem categoria" (ver a
 *     nota CATEGORIA OBRIGATÓRIA em app/models/movimento.py, no backend).
 *   - "saldo_apos" — o saldo da conta imediatamente depois deste
 *     movimento, calculado no servidor sobre o histórico COMPLETO da
 *     conta (nunca só sobre a janela/filtro pedidos). Só vem preenchido
 *     em listarMovimentos (GET /movimentos); nos restantes pedidos
 *     (criar, obter um só, editar) fica null, porque nada os usa para o
 *     mostrar.
 */
export type Movimento = {
  id: string
  conta_id: string
  categoria_id: string
  data: string
  descricao: string
  valor: string
  created_at: string
  updated_at: string
  saldo_apos: string | null
}

/** Campos aceites ao criar ou editar um movimento (schemas MovimentoCriar
 *  / MovimentoEditar — a mesma forma nas duas, ao contrário de Conta). */
export type MovimentoDados = {
  conta_id: string
  categoria_id: string
  data: string
  descricao: string
  valor: string
}

/**
 * O cursor da PÁGINA SEGUINTE de listarMovimentos — os três campos pelos
 * quais a lista está ordenada no backend (ver a nota PAGINAÇÃO POR CURSOR
 * em app/routers/movimentos.py), tirados do ÚLTIMO movimento da página já
 * recebida. Pedir a próxima página é repetir o pedido com este cursor.
 */
export type CursorMovimentos = {
  data: string
  criadoEm: string
  id: string
}

/**
 * Os parâmetros aceites por listarMovimentos — espelham exactamente os da
 * rota GET /movimentos no backend (ver a nota FILTROS COMO PARÂMETROS, no
 * mesmo ficheiro): "contaId" filtra para uma única conta (ainda sem uso na
 * interface — fica para quando existir "ver movimentos desta conta");
 * "contas"/"categorias" filtram por uma ou mais contas/categorias;
 * "tipo", "de"/"ate" e "pesquisa" tal como na folha de filtros e no campo
 * de procura da página. "cursor", omitido, pede a PRIMEIRA página; "limite"
 * tem a mesma omissão por omissão do backend (30) quando não indicado.
 */
export type ListarMovimentosOpcoes = {
  contaId?: string
  contas?: string[]
  categorias?: string[]
  tipo?: 'entrada' | 'saida' | null
  de?: string | null
  ate?: string | null
  pesquisa?: string
  cursor?: CursorMovimentos | null
  limite?: number
}

/**
 * Lista uma JANELA de movimentos do utilizador (todas as suas contas), por
 * data mais recente primeiro — nunca o histórico completo de uma vez (ver
 * a nota PAGINAÇÃO POR CURSOR em app/routers/movimentos.py, no backend).
 * Sem "cursor", devolve a primeira página; com ele, continua a partir do
 * último movimento da página anterior.
 */
export function listarMovimentos(opcoes: ListarMovimentosOpcoes = {}): Promise<Movimento[]> {
  const params = new URLSearchParams()
  if (opcoes.contaId) params.set('conta_id', opcoes.contaId)
  if (opcoes.contas && opcoes.contas.length > 0) params.set('contas', opcoes.contas.join(','))
  if (opcoes.categorias && opcoes.categorias.length > 0) {
    params.set('categorias', opcoes.categorias.join(','))
  }
  if (opcoes.tipo) params.set('tipo', opcoes.tipo)
  if (opcoes.de) params.set('de', opcoes.de)
  if (opcoes.ate) params.set('ate', opcoes.ate)
  if (opcoes.pesquisa) params.set('pesquisa', opcoes.pesquisa)
  if (opcoes.cursor) {
    params.set('antes_data', opcoes.cursor.data)
    params.set('antes_criado_em', opcoes.cursor.criadoEm)
    params.set('antes_id', opcoes.cursor.id)
  }
  if (opcoes.limite) params.set('limite', String(opcoes.limite))

  const query = params.toString()
  return pedido<Movimento[]>(`/movimentos${query ? `?${query}` : ''}`)
}

export function obterMovimento(id: string): Promise<Movimento> {
  return pedido<Movimento>(`/movimentos/${id}`)
}

export function criarMovimento(dados: MovimentoDados): Promise<Movimento> {
  return pedido<Movimento>('/movimentos', { method: 'POST', body: JSON.stringify(dados) })
}

export function editarMovimento(id: string, dados: MovimentoDados): Promise<Movimento> {
  return pedido<Movimento>(`/movimentos/${id}`, { method: 'PATCH', body: JSON.stringify(dados) })
}

export function apagarMovimento(id: string): Promise<void> {
  return pedido<void>(`/movimentos/${id}`, { method: 'DELETE' })
}

/**
 * Eliminar vários movimentos de uma vez — o "Eliminar" do modo de seleção
 * múltipla da lista. Atómico do lado do servidor: ou apaga todos, ou (um
 * id inválido) nenhum.
 */
export function eliminarMovimentosEmLote(ids: string[]): Promise<void> {
  return pedido<void>('/movimentos/eliminar-em-lote', {
    method: 'POST',
    body: JSON.stringify({ ids }),
  })
}

/**
 * Recategorizar vários movimentos de uma vez — o "Categorizar" do modo de
 * seleção múltipla. Atómico: se a categoria não for coerente com a
 * direção de algum dos movimentos do lote, nenhum é alterado.
 */
export function recategorizarMovimentosEmLote(ids: string[], categoriaId: string): Promise<void> {
  return pedido<void>('/movimentos/recategorizar-em-lote', {
    method: 'POST',
    body: JSON.stringify({ ids, categoria_id: categoriaId }),
  })
}
